migrator/src/main/java/com/google/phonenumbers/migrator/MigrationJob.java

```diff
@@ -73,29 +79,31 @@ public PhoneRegion getRegionCode() {
 
   /**
    * Returns the sub range of numbers within numberRange that can be migrated using any recipe from
-   * the {@link CsvTable} recipesTable that matches the specified BCP-47 region code. This method will
-   * not perform migrations and as a result, the validity of migrations using the given recipesTable
-   * cannot be verified.
+   * the {@link CsvTable} recipesTable that matches the specified BCP-47 region code. This method
+   * will not perform migrations and as a result, the validity of migrations using the given
+   * recipesTable cannot be verified.
    */
-  public RangeTree getAllMigratableNumbers() {
+  public Stream<DigitSequence> getAllMigratableNumbers() {
     return RecipesTableSchema.toRangeTable(recipesTable)
-        .getRanges(RecipesTableSchema.REGION_CODE, regionCode)
-        .intersect(getNumberRange());
+        .getRanges(RecipesTableSchema.REGION_CODE, regionCode).asRangeSet()
+        .intersection(getNumberRange()).asRanges().stream().map(Range::lowerEndpoint);
```

Comment:
So you have a RangeTree (the intersection) and you want to stream digit sequences?
This code won't do that I'm afraid (if you have 3 numbers next to each other you only get the first).

The reason it is hard in the API to do this is because most of the time we don't want to encourage an easy way to do this, because these trees have something like 10^20 elements "in them", so iterating one of those would just lock up and application. Even if what you had done worked, I think we can both agree that it feels rather unsatisfying; it just feels like "too much mucking around".

This is the sort of 6th sense you want to develop when writing your own code or reading other peoples. Even if code does exactly what it needs to do, there can still be clear signs that it's "wrong" (in the sense that it should not have to be as complex as it is).

So, let's think about other ways to represent the operation:

You have two inputs, the "number range" (the set of input numbers) and the range table to check it against.
One has a very small set of elements, one has a potentially massive set.
And the result we want is a filtered version of the small set.

In cases like this, it's often worth re-framing the operation to start with the smaller input. This bounds complexity naturally. See if you can figure out a different, much simpler way to do the operation (lets say assuming that getNumberRange() returned a Stream<DigitSequence> as well).